File: haskell/ColorRefineBarebones.hs
{-# LANGUAGE BangPatterns #-}
-- Color refinement (1-WL) benchmark — barebones (mutable unboxed vectors, unsafeRead/unsafeWrite)
-- Integer widths: Word32 for node IDs/offsets/colors, Word64 for hashes, Int64 for checksum.
module Main where

import qualified Data.ByteString.Char8 as BS
import qualified Data.Vector.Unboxed.Mutable as MV
import qualified Data.HashMap.Strict as HM
import Data.Int (Int64)
import Data.Word (Word32, Word64)
import Data.Bits (xor)
import System.Environment (getArgs)
import System.Clock (getTime, Clock(Monotonic), toNanoSecs)
import Numeric (showFFloat)

readInt :: BS.ByteString -> (Int, BS.ByteString)
readInt !bs = case BS.readInt (BS.dropWhile (< '0') bs) of
    Just (!n, !rest) -> (n, rest)
    Nothing -> error "parse error"

i2w :: Int -> Word32
i2w = fromIntegral
{-# INLINE i2w #-}

w2i :: Word32 -> Int
w2i = fromIntegral
{-# INLINE w2i #-}

main :: IO ()
main = do
    args <- getArgs
    case args of
      [file] -> run file
      _ -> putStrLn "Usage: color_refine_barebones <graph_file>"

run :: FilePath -> IO ()
run file = do
    -- ── Read ─────────────────────────────────────────────────────────────
    t0 <- getTime Monotonic
    contents <- BS.readFile file

    let (!n, r0) = readInt contents
        (!m, r1) = readInt r0

    -- Count degrees
    deg <- MV.replicate n (0 :: Word32)
    let readEdges !buf !rem !i
          | i >= m    = return rem
          | otherwise = do
              let (!u, r2) = readInt rem
                  (!v, r3) = readInt r2
              MV.unsafeWrite buf (2*i) (i2w u)
              MV.unsafeWrite buf (2*i+1) (i2w v)
              MV.unsafeRead deg u >>= MV.unsafeWrite deg u . (+1)
              MV.unsafeRead deg v >>= MV.unsafeWrite deg v . (+1)
              readEdges buf r3 (i+1)

    edgeBuf <- MV.new (2*m) :: IO (MV.IOVector Word32)
    _ <- readEdges edgeBuf r1 0

    -- Build CSR
    offset <- MV.new (n+1) :: IO (MV.IOVector Word32)
    MV.unsafeWrite offset 0 0
    let buildOffset !i
          | i >= n    = return ()
          | otherwise = do
              !prev <- MV.unsafeRead offset i
              !d    <- MV.unsafeRead deg i
              MV.unsafeWrite offset (i+1) (prev + d)
              buildOffset (i+1)
    buildOffset 0

    total <- MV.unsafeRead offset n
    adj <- MV.new (w2i total) :: IO (MV.IOVector Word32)
    pos <- MV.replicate n (0 :: Word32)

    let fillAdj !i
          | i >= m    = return ()
          | otherwise = do
              !u <- MV.unsafeRead edgeBuf (2*i)
              !v <- MV.unsafeRead edgeBuf (2*i+1)
              let !ui = w2i u
                  !vi = w2i v
              !ou <- MV.unsafeRead offset ui
              !pu <- MV.unsafeRead pos ui
              MV.unsafeWrite adj (w2i (ou+pu)) v
              MV.unsafeWrite pos ui (pu+1)
              !ov <- MV.unsafeRead offset vi
              !pv <- MV.unsafeRead pos vi
              MV.unsafeWrite adj (w2i (ov+pv)) u
              MV.unsafeWrite pos vi (pv+1)
              fillAdj (i+1)
    fillAdj 0

    t1 <- getTime Monotonic
    let readMs = fromIntegral (toNanoSecs t1 - toNanoSecs t0) / 1e6 :: Double

    -- ── Compute ──────────────────────────────────────────────────────────
    t2 <- getTime Monotonic

    color <- MV.replicate n (0 :: Word32)
    newColor <- MV.new n :: IO (MV.IOVector Word32)
    sigHash <- MV.new n :: IO (MV.IOVector Word64)

    -- Find max degree for reusable buffer
    let findMaxDeg !v !mx
          | v >= n    = return mx
          | otherwise = do
              !lo <- MV.unsafeRead offset v
              !hi <- MV.unsafeRead offset (v+1)
              let !d = w2i (hi - lo)
              findMaxDeg (v+1) (max mx d)
    !maxDeg <- findMaxDeg 0 0
    nbuf <- MV.new maxDeg :: IO (MV.IOVector Word32)

    let refine !round !prevRounds !prevColors
          | round >= n = return (prevRounds, prevColors)
          | otherwise = do
              -- Build signature hash for each vertex
              let hashVertex !v
                    | v >= n    = return ()
                    | otherwise = do
                        !lo <- MV.unsafeRead offset v
                        !hi <- MV.unsafeRead offset (v+1)
                        let !degV = w2i (hi - lo)
                        -- Collect neighbor colors
                        let collectNbr !j
                              | j >= degV = return ()
                              | otherwise = do
                                  !w <- MV.unsafeRead adj (w2i lo + j)
                                  !c <- MV.unsafeRead color (w2i w)
                                  MV.unsafeWrite nbuf j c
                                  collectNbr (j+1)
                        collectNbr 0
                        -- Sort neighbor colors in-place (insertion sort)
                        let insertSort !i
                              | i >= degV = return ()
                              | otherwise = do
                                  !key <- MV.unsafeRead nbuf i
                                  let shift !j
                                        | j <= 0    = MV.unsafeWrite nbuf j key >> insertSort (i+1)
                                        | otherwise = do
                                            !prev <- MV.unsafeRead nbuf (j-1)
                                            if prev > key
                                              then MV.unsafeWrite nbuf j prev >> shift (j-1)
                                              else MV.unsafeWrite nbuf j key >> insertSort (i+1)
                                  shift i
                        insertSort 1
                        -- Hash
                        !cv <- MV.unsafeRead color v
                        let !h0' = fromIntegral cv * 1000003 :: Word64
                            !h0  = (h0' `xor` (fromIntegral degV * 2654435761)) * 1000003
                        let hashNbr !j !h
                              | j >= degV = return h
                              | otherwise = do
                                  !nc <- MV.unsafeRead nbuf j
                                  let !h' = (h `xor` (fromIntegral nc * 2654435761)) * 1000003
                                  hashNbr (j+1) h'
                        !hv <- hashNbr 0 h0
                        MV.unsafeWrite sigHash v hv
                        hashVertex (v+1)
              hashVertex 0

              -- Map hashes to consecutive colors
              let buildMapping !v !nextId !mp
                    | v >= n    = return (nextId, mp)
                    | otherwise = do
                        !h <- MV.unsafeRead sigHash v
                        case HM.lookup h mp of
                          Just !cid -> do
                            MV.unsafeWrite newColor v cid
                            buildMapping (v+1) nextId mp
                          Nothing -> do
                            MV.unsafeWrite newColor v nextId
                            buildMapping (v+1) (nextId+1) (HM.insert h nextId mp)
              (!nColors, _) <- buildMapping 0 0 HM.empty

              -- Check stability
              let checkStable !v
                    | v >= n    = return True
                    | otherwise = do
                        !c <- MV.unsafeRead color v
                        !nc <- MV.unsafeRead newColor v
                        if c /= nc then return False
                        else checkStable (v+1)
              !stable <- checkStable 0

              let !thisRound = round + 1
              if stable
                then return (thisRound, nColors)
                else do
                  -- Copy newColor -> color
                  MV.unsafeCopy color newColor
                  refine (round+1) thisRound nColors

    (!rounds, !numColors) <- refine 0 0 0

    -- Compute checksum
    let sumColors !v !s
          | v >= n    = return s
          | otherwise = do
              !c <- MV.unsafeRead color v
              sumColors (v+1) (s + fromIntegral c)
    !checksum <- sumColors 0 (0 :: Int64)

    t3 <- getTime Monotonic
    let computeMs = fromIntegral (toNanoSecs t3 - toNanoSecs t2) / 1e6 :: Double

    putStrLn $ "read=" ++ showFFloat (Just 1) readMs "" ++ "ms compute="
              ++ showFFloat (Just 1) computeMs "" ++ "ms rounds=" ++ show rounds
              ++ " colors=" ++ show numColors ++ " checksum=" ++ show checksum

File: cpp/color_refine_barebones.cpp
// Color refinement (1-WL) benchmark — barebones (raw arrays, manual hashing)
// Input: graph file (n m, then m edges u v)
// Output: read=Xms compute=Yms rounds=R colors=C checksum=K
#include <cstdio>
#include <cstdlib>
#include <cstdint>
#include <cstring>
#include <chrono>
#include <algorithm>
#include <unordered_map>

using Clock = std::chrono::steady_clock;

static double ms_since(Clock::time_point t0) {
    return std::chrono::duration<double, std::milli>(Clock::now() - t0).count();
}

int main(int argc, char **argv) {
    if (argc < 2) { fprintf(stderr, "Usage: %s <graph_file>\n", argv[0]); return 1; }

    // ── Read ─────────────────────────────────────────────────────────────────
    auto t0 = Clock::now();

    FILE *f = fopen(argv[1], "r");
    if (!f) { perror("fopen"); return 1; }

    uint32_t n, m;
    fscanf(f, "%u %u", &n, &m);

    // CSR-style adjacency: offsets + flat neighbor array
    // First pass: count degrees
    uint32_t *deg = (uint32_t *)calloc(n, sizeof(uint32_t));
    uint32_t *eu = (uint32_t *)malloc(m * sizeof(uint32_t));
    uint32_t *ev = (uint32_t *)malloc(m * sizeof(uint32_t));
    for (uint32_t i = 0; i < m; i++) {
        fscanf(f, "%u %u", &eu[i], &ev[i]);
        deg[eu[i]]++;
        deg[ev[i]]++;
    }
    fclose(f);

    // Build adjacency arrays
    uint32_t *offset = (uint32_t *)malloc((n + 1) * sizeof(uint32_t));
    offset[0] = 0;
    for (uint32_t i = 0; i < n; i++) offset[i + 1] = offset[i] + deg[i];
    uint32_t total_edges = offset[n];
    uint32_t *adj = (uint32_t *)malloc(total_edges * sizeof(uint32_t));
    uint32_t *pos = (uint32_t *)calloc(n, sizeof(uint32_t));
    for (uint32_t i = 0; i < m; i++) {
        uint32_t u = eu[i], v = ev[i];
        adj[offset[u] + pos[u]++] = v;
        adj[offset[v] + pos[v]++] = u;
    }
    free(eu); free(ev); free(deg); free(pos);

    double read_ms = ms_since(t0);

    // ── Compute ──────────────────────────────────────────────────────────────
    auto t1 = Clock::now();

    uint32_t *color = (uint32_t *)calloc(n, sizeof(uint32_t));
    uint32_t *new_color = (uint32_t *)malloc(n * sizeof(uint32_t));
    uint64_t *sig_hash = (uint64_t *)malloc(n * sizeof(uint64_t));
    // Reusable buffer for neighbor colors
    uint32_t max_deg = 0;
    for (uint32_t v = 0; v < n; v++) {
        uint32_t d = offset[v + 1] - offset[v];
        if (d > max_deg) max_deg = d;
    }
    uint32_t *nbuf = (uint32_t *)malloc(max_deg * sizeof(uint32_t));

    uint32_t rounds = 0;
    uint32_t num_colors = 0;

    for (uint32_t round = 0; round < n; round++) {
        // Build signature hash for each vertex
        for (uint32_t v = 0; v < n; v++) {
            uint32_t lo = offset[v], hi = offset[v + 1];
            uint32_t deg_v = hi - lo;
            for (uint32_t i = 0; i < deg_v; i++)
                nbuf[i] = color[adj[lo + i]];
            std::sort(nbuf, nbuf + deg_v);

            uint64_t h = (uint64_t)color[v] * 1000003ULL;
            h = h ^ ((uint64_t)deg_v * 2654435761ULL);
            h = h * 1000003ULL;
            for (uint32_t i = 0; i < deg_v; i++) {
                h = h ^ ((uint64_t)nbuf[i] * 2654435761ULL);
                h = h * 1000003ULL;
            }
            sig_hash[v] = h;
        }

        // Map hashes to consecutive colors
        std::unordered_map<uint64_t, uint32_t> mapping;
        uint32_t next_id = 0;
        for (uint32_t v = 0; v < n; v++) {
            auto it = mapping.find(sig_hash[v]);
            if (it == mapping.end()) {
                mapping[sig_hash[v]] = next_id;
                new_color[v] = next_id;
                next_id++;
            } else {
                new_color[v] = it->second;
            }
        }

        rounds = round + 1;
        num_colors = next_id;

        // Check stability
        bool stable = true;
        for (uint32_t v = 0; v < n; v++) {
            if (new_color[v] != color[v]) { stable = false; break; }
        }
        if (stable) break;

        memcpy(color, new_color, n * sizeof(uint32_t));
    }

    // Compute checksum
    int64_t checksum = 0;
    for (uint32_t v = 0; v < n; v++) checksum += color[v];

    double compute_ms = ms_since(t1);

    // ── Output ───────────────────────────────────────────────────────────────
    printf("read=%.1fms compute=%.1fms rounds=%u colors=%u checksum=%lld\n",
           read_ms, compute_ms, rounds, num_colors, (long long)checksum);

    free(color); free(new_color); free(sig_hash); free(nbuf);
    free(adj); free(offset);
    return 0;
}
